components/kyma-environment-broker/internal/process/provisioning/connectivity_bind.go

```diff
@@ -106,82 +86,24 @@ func (s *ConnectivityBindStep) Run(operation internal.ProvisioningOperation, log
 		operation = op
 	} else {
 		// get the credentials from encrypted string in operation.Connectivity.Instance.
```

Comment:
I'm afraid we will never come into this place. Returning  `operation, 0, nil` in line 98 after first pass will make Executor to think this step is completed and the move to the next step of processing the operation. See line 94 from the code of `components/kyma-environment-broker/internal/process/upgrade_kyma/manager.go`

```
for _, weightStep := range m.sortWeight() {
		steps := m.steps[weightStep]
		for _, step := range steps {
			logStep := logOperation.WithField("step", step.Name())
			logStep.Infof("Start step")

			operation, when, err = m.runStep(step, operation, logStep)
			if err != nil {
				logStep.Errorf("Process operation failed: %s", err)
				return 0, err
			}
			if operation.IsFinished() {
				logStep.Infof("Operation %q got status %s. Process finished.", operation.Operation.ID, operation.State)
				return 0, nil
			}
// this condition will be true after first pass and we will move to the next step 
			if when == 0 {
				logStep.Info("Process operation successful")
				continue
			}

			logStep.Infof("Process operation will be repeated in %s ...", when)
			return when, nil
		}
	}
````

Reply:
The second thing is we could maybe simplify it and update operation in one step just after extracting connectivity overrides. This will allow us to do not Encrypt/Decrypt credentials between two runs which is strange.


So the whole solution could looks like following:

```		
		respBinding, err := smCli.Bind(operation.Connectivity.Instance.InstanceKey(), operation.Connectivity.BindingID, nil, false)
		if err != nil {
			return s.handleError(operation, err, log, fmt.Sprintf("Bind() call failed"))
		}
		// get overrides
		connectivityOverrides, err = connectivity_bind.GetConnectivityCredentials(respBinding.Binding)
		if err != nil {
			return s.handleError(operation, err, log, fmt.Sprintf("unable to load config"))
		}

                 connectivityOverrides, err = connectivity_bind.GetConnectivityCredentials(respBinding.Binding)
		if err != nil {
			return s.handleError(operation, err, log, fmt.Sprintf("unable to load config"))
		}

     	// save the status
		op, retry := s.operationManager.UpdateOperation(operation, func(operation *internal.ProvisioningOperation) {
			applyOverrides(connectivityOverrides, *operation)
			operation.Connectivity.Instance.Provisioned = true
			operation.Connectivity.Instance.ProvisioningTriggered = false
		}, log)
		if retry > 0 {
			log.Errorf("unable to update operation")
			return operation, time.Second, nil
		}
		operation = op

	}

	log.Debugf("Got Connectivity Service credentials from the binding.")

	return operation, 0, nil
```